fem/bilininteg_dgtrace.cpp

```diff
@@ -0,0 +1,1092 @@
+            for (int q = 0; q < nq; ++q)
+            {
+               // Convert to lexicographic ordering
+               int iq = ToLexOrdering(dim, face_id, quad1D, q);
+               C(iq,f_ind) = rho->Eval(T, ir->IntPoint(q));
+            }
+            f_ind++;
+         }
+      }
+      MFEM_VERIFY(f_ind==nf, "Incorrect number of faces.");
+   }
+   Vector vel;
+   if (VectorConstantCoefficient *c_u = dynamic_cast<VectorConstantCoefficient*>
+                                        (u))
+   {
+      vel = c_u->GetVec();
+   }
+   else
+   {
+      vel.SetSize(dim * nq * nf);
+      auto C = Reshape(vel.HostWrite(), dim, nq, nf);
+      Vector Vq(dim);
+      int f_ind = 0;
+      for (int f = 0; f < fes.GetNF(); ++f)
+      {
+         int e1, e2;
+         int inf1, inf2;
+         fes.GetMesh()->GetFaceElements(f, &e1, &e2);
+         fes.GetMesh()->GetFaceInfos(f, &inf1, &inf2);
+         int face_id = inf1 / 64;
+         if ((type==FaceType::Interior && (e2>=0 || (e2<0 && inf2>=0))) ||
+             (type==FaceType::Boundary && e2<0 && inf2<0) )
+         {
+            ElementTransformation& T = *fes.GetMesh()->GetFaceTransformation(f);
+            for (int q = 0; q < nq; ++q)
+            {
+               // Convert to lexicographic ordering
+               int iq = ToLexOrdering(dim, face_id, quad1D, q);
+               u->Eval(Vq, T, ir->IntPoint(q));
+               for (int i = 0; i < dim; ++i)
+               {
+                  C(i,iq,f_ind) = Vq(i);
+               }
+            }
+            f_ind++;
+         }
+      }
+      MFEM_VERIFY(f_ind==nf, "Incorrect number of faces.");
+   }
+   PADGTraceSetup(dim, dofs1D, quad1D, nf, ir->GetWeights(),
+                  geom->detJ, geom->normal, r, vel,
+                  alpha, beta, pa_data);
+}
+
+void DGTraceIntegrator::AssemblePAInteriorFaces(const FiniteElementSpace& fes)
+{
+   SetupPA(fes, FaceType::Interior);
+}
+
+void DGTraceIntegrator::AssemblePABoundaryFaces(const FiniteElementSpace& fes)
+{
+   SetupPA(fes, FaceType::Boundary);
+}
+
+// PA DGTrace Apply 2D kernel for Gauss-Lobatto/Bernstein
+template<int T_D1D = 0, int T_Q1D = 0> static
+void PADGTraceApply2D(const int NF,
+                      const Array<double> &b,
+                      const Array<double> &bt,
+                      const Vector &_op,
+                      const Vector &_x,
+                      Vector &_y,
+                      const int d1d = 0,
+                      const int q1d = 0)
+{
+   const int VDIM = 1;
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   MFEM_VERIFY(D1D <= MAX_D1D, "");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   auto B = Reshape(b.Read(), Q1D, D1D);
+   auto Bt = Reshape(bt.Read(), D1D, Q1D);
+   auto op = Reshape(_op.Read(), Q1D, 2, 2, NF);
+   auto x = Reshape(_x.Read(), D1D, VDIM, 2, NF);
+   auto y = Reshape(_y.ReadWrite(), D1D, VDIM, 2, NF);
```

Comment:
what does the 2 in line 268 correspond to?

Reply:
`Elem1` or `Elem2` on the face.

Reply:
How do you account for an element being on the boundary?

Reply:
Interior and boundary faces are treated independently since you can't assume the same Integrators.

Reply:
You specify to `FaceRestriction` if you're treating boundary or interior faces.